float32[] matrix_row